### PR TITLE
impl(pubsub): use publisher tracing connection when otel is enabled

### PR DESCRIPTION
**Pull request by alevenberg**
Use the MakePublisherTracingConnection if the Open Telemetry tracing option is set

Closes #[12525](https://github.com/googleapis/google-cloud-cpp/issues/12525)

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12634)

### CODE REVIEW

File: google/cloud/pubsub/publisher_connection_test.cc [hunk 1]
```diff
@@ -334,6 +340,60 @@ TEST(PublisherConnectionTest, HandleTransientEnabledRetry) {
...
+using ::testing::Contains;
+using ::testing::IsEmpty;
+
+TEST(MakePublisherConnectionTest, TracingEnabled) {
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_shared<pubsub_testing::MockPublisherStub>();
+  Topic const topic("test-project", "test-topic");
+  EXPECT_CALL(*mock, AsyncPublish)
+      .WillOnce([&](google::cloud::CompletionQueue&, auto,
+                    google::pubsub::v1::PublishRequest const& request) {
```
Comment (dbolduc) [inline google/cloud/pubsub/publisher_connection_test.cc:357]: compiler is mad at you for naming `request` here and below

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/publisher_connection_test.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/publisher_connection_test.cc]: the compiler got smarter.... older code does it and doesn't get yelled at 😢 

fixed.

File: google/cloud/pubsub/publisher_connection.cc [hunk 2]
```diff
@@ -59,14 +61,21 @@ std::shared_ptr<pubsub::PublisherConnection> ConnectionFromDecoratedStub(
     }
     return pubsub_internal::RejectsWithOrderingKey::Create(
         pubsub_internal::BatchingPublisherConnection::Create(
-            std::move(topic), opts, {}, std::move(sink), std::move(cq)));
+            topic, opts, {}, std::move(sink), std::move(cq)));
   };
+  auto enable_open_telemetry = google::cloud::internal::TracingEnabled(opts);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/publisher_connection.cc]: done

File: google/cloud/pubsub/publisher_connection_test.cc [hunk 3]
```diff
@@ -334,6 +340,60 @@ TEST(PublisherConnectionTest, HandleTransientEnabledRetry) {
...
+  auto publisher =
+      MakeTestPublisherConnection(topic, mock, DisableTracing(Options{}));
+
+  auto response =
+      publisher->Publish({MessageBuilder{}.SetData("test-data-0").Build()})
+          .get();
+
+  EXPECT_THAT(span_catcher->GetSpans(), IsEmpty());
+}
+
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/publisher_connection_test.cc]: done

File: google/cloud/pubsub/publisher_connection_test.cc [hunk 4]
```diff
@@ -23,6 +24,11 @@
 #include "google/cloud/testing_util/status_matchers.h"
 #include "google/cloud/testing_util/validate_metadata.h"
 #include <gmock/gmock.h>
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/opentelemetry_options.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/publisher_connection_test.cc:29]: Done

**Pull request merged**